### PR TITLE
Tripal 4 chado_unabbreviate_infraspecific_rank (port tripal 3 pull 1244)

### LINKED ISSUE #1485: Port pull #1244 to Tripal 4
# Core Tripal 4 Development Task

## Branch Name: tv4g0-issue1485-port-pull-1244

Port pull #1244 to tripal 4, isue #1243.
This pull added function `chado_unabbreviate_infraspecific_rank($rank)` which is an inverse of existing function `chado_abbreviate_infraspecific_rank($rank)`.

### PR DESCRIPTION
**Pull request by dsenalik**
# Tripal 4 Core Dev Task

## Description
This ports Tripal 3 pull request #1244 to Tripal 4.
This added a new api function `chado_unabbreviate_infraspecific_rank()` which is an inverse of the existing `chado_abbreviate_infraspecific_rank()`

This also adds automated testing for most of the organism API functions:
- `function chado_get_organism()`
- `function chado_get_organism_scientific_name()`
- `function chado_get_organism_select_options()`
- `function chado_abbreviate_infraspecific_rank()`
- `function chado_unabbreviate_infraspecific_rank()` (the function added here)

Not tested yet:
- `function chado_get_organism_image_url()`
- `function chado_autocomplete_organism()`

## Testing?
You can run this in /devel/php
```
$testarray = ['subsp', 'ssp', 'var', 'subvar', 'cv', 'f', 'subf', 'bogus'];
foreach ($testarray as $name) {
  print "\"$name\" becomes \"".chado_unabbreviate_infraspecific_rank($name)."\"\n";
  $name .= '.';
  print "\"$name\" becomes \"".chado_unabbreviate_infraspecific_rank($name)."\"\n";
}
```
Output
```
"subsp" becomes "subspecies"
"subsp." becomes "subspecies"
"ssp" becomes "subspecies"
"ssp." becomes "subspecies"
"var" becomes "varietas"
"var." becomes "varietas"
"subvar" becomes "subvarietas"
"subvar." becomes "subvarietas"
"cv" becomes "cultivar"
"cv." becomes "cultivar"
"f" becomes "forma"
"f." becomes "forma"
"subf" becomes "subforma"
"subf." becomes "subforma"
"bogus" becomes "bogus"
"bogus." becomes "bogus."
```

### CODE REVIEW

**Comment by laceysanderson** [comment]
Note: Code coverage is 100% on this PR because the file ending is incorrect -it should be `.php`. That said, I've added an issue to change this for all api files in tripal_chado and that particular piece of information does not impact this PR's validity. Although it would be nice if it had some automated tests ;-p

**Comment by dsenalik** [comment]
I have created tests for most of the organism API functions, the new function from this pull request should have true 100% coverage. I have not renamed from `.inc` to `.php` in this pull request, that should be done for all files together for issue #1502
Ready for review, especially for my automated tests, did I do them right?

**Comment by dsenalik** [comment]
Because of the significant overhead of running a test function, I have combined all organism API tests into a single test function `test_chado_organism_api_functions()`

File: tripal_chado/tests/src/Functional/api/ChadoOrganismAPITest.php [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+   * Modules to enable.
+   * @var array
+   */
+  protected static $modules = ['tripal', 'tripal_chado'];
+
+  /**
+   * Schema to do testing out of.
+   * @var string
+   */
+  protected static $schemaName = 'testchado';
```
Comment (laceysanderson) [inline tripal_chado/tests/src/Functional/api/ChadoOrganismAPITest.php]: We are phasing out using the `testchado` schema which you would have no way of knowing 🙈 I'll add in here the new way as suggestions (in two parts).

```suggestion
  protected $schemaName;
```

File: tripal_chado/tests/src/Functional/api/ChadoOrganismAPITest.php [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+   *   chado_abbreviate_infraspecific_rank()
+   *   chado_unabbreviate_infraspecific_rank()
+   *
+   * @group tripal-chado
+   * @group chado-organism
+   */
+  public function test_chado_organism_api_functions() {
+
+    putenv('TRIPAL_SUPPRESS_ERRORS=TRUE');
+
```
Comment (laceysanderson) [inline tripal_chado/tests/src/Functional/api/ChadoOrganismAPITest.php:50]:
```suggestion

// Create a Test Chado schema which contains the expected cvterms added during the prepare step.
self::$schemaName = $this->createTestSchema(ChadoTestBrowserBase::PREPARE_TEST_CHADO);
```

File: tripal_chado/tests/src/Functional/api/ChadoOrganismAPITest.php [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    // Lookup cvterm_id for 'subspecies'
+    $cvterm = chado_get_cvterm(['name' => 'subspecies'], [], self::$schemaName);
+    $this->assertNotNull($cvterm, 'test_chado_organism_api_functions() unable to retrieve cvterm for subspecies');
```
Comment (laceysanderson) [inline tripal_chado/tests/src/Functional/api/ChadoOrganismAPITest.php]: The failed test result already tells you the test function in the stack trace so I like to make this message specific to the part of the test I'm currently running. For example,

```suggestion
    $this->assertNotNull($cvterm, 'Unable to retrieve cvterm for subspecies using chado_get_cvterm when testing the Organism API.');
```

But this is just a preference change and yours is good too :-) So feel free to ignore.

Comment (laceysanderson) [inline tripal_chado/tests/src/Functional/api/ChadoOrganismAPITest.php:50]: The failed test result already tells you the test function in the stack trace so I like to make this message specific to the part of the test I'm currently running. For example,

```suggestion
    $this->assertNotNull($cvterm, 'Unable to retrieve cvterm for subspecies using chado_get_cvterm when testing the Organism API.');
```

But this is just a preference change and yours is good too :-) So feel free to ignore.

**Review by laceysanderson — COMMENTED** [review]
✅ Manual Testing looked perfect 👍 
Screenshot 2023-05-10 at 9 53 32 AM

✅ Code review of API functions is good :-)

✅ Code review of testings is also good as is although there are a few suggestions. I would say the only important suggestion is regarding `testchado` which we are trying to phase out 🙃 but your usage is not wrong and we didn't document thing to phase it out so 🙈

File: tripal_chado/tests/src/Functional/api/ChadoOrganismAPITest.php (hunk 1 above)
Comment (dsenalik) [reply tripal_chado/tests/src/Functional/api/ChadoOrganismAPITest.php]: Review suggestions appreciated and incorporated

**Pull request merged**